pkg/datastore/test/pagination.go

```diff
@@ -0,0 +1,315 @@
+	}
+
+	rawDS, err := tester.New(0, veryLargeGCInterval, veryLargeGCWindow, 1)
+	require.NoError(t, err)
+
+	ds, rev := testfixtures.StandardDatastoreWithData(rawDS, require.New(t))
+	tRequire := testfixtures.TupleChecker{Require: require.New(t), DS: ds}
+
+	for _, tc := range testCases {
+		expected := sortedStandardData(tc.objectType, options.ByResource)
+		for limit := 1; limit <= len(expected); limit++ {
+			testLimit := uint64(limit)
+
+			t.Run(fmt.Sprintf("%s-%d", tc.objectType, tc.ordering), func(t *testing.T) {
+				require := require.New(t)
+				ctx := context.Background()
+
+				expected := sortedStandardData(tc.objectType, tc.ordering)
+
+				foreachTxType(ctx, ds, rev, func(reader datastore.Reader) {
+					iter, err := reader.QueryRelationships(ctx, datastore.RelationshipsFilter{
+						ResourceType: tc.objectType,
+					}, options.WithSort(tc.ordering), options.WithLimit(&testLimit))
+
+					require.NoError(err)
+					defer iter.Close()
+
+					cursor, err := iter.Cursor()
+					require.Empty(cursor)
+					require.ErrorIs(err, datastore.ErrCursorEmpty)
+
+					tRequire.VerifyOrderedIteratorResults(iter, expected[0:limit]...)
+
+					cursor, err = iter.Cursor()
+					if len(expected) > 0 {
+						require.NotEmpty(cursor)
+						require.NoError(err)
+					} else {
+						require.Empty(cursor)
+						require.ErrorIs(err, datastore.ErrCursorEmpty)
+					}
+				})
+			})
+		}
+	}
+}
+
+func ResumeTest(t *testing.T, tester DatastoreTester) {
+	testCases := []struct {
+		objectType string
+		ordering   options.SortOrder
+	}{
+		{testfixtures.DocumentNS.Name, options.ByResource},
+		{testfixtures.DocumentNS.Name, options.BySubject},
+		{testfixtures.FolderNS.Name, options.ByResource},
+		{testfixtures.FolderNS.Name, options.BySubject},
+		{testfixtures.UserNS.Name, options.ByResource},
+		{testfixtures.UserNS.Name, options.BySubject},
+	}
+
+	rawDS, err := tester.New(0, veryLargeGCInterval, veryLargeGCWindow, 1)
+	require.NoError(t, err)
+
+	ds, rev := testfixtures.StandardDatastoreWithData(rawDS, require.New(t))
+	tRequire := testfixtures.TupleChecker{Require: require.New(t), DS: ds}
+
+	for _, tc := range testCases {
+		expected := sortedStandardData(tc.objectType, options.ByResource)
+		for batchSize := 1; batchSize <= len(expected); batchSize++ {
+			testLimit := uint64(batchSize)
+
+			t.Run(fmt.Sprintf("%s-%d-batches-%d", tc.objectType, tc.ordering, batchSize), func(t *testing.T) {
+				require := require.New(t)
+				ctx := context.Background()
+
+				expected := sortedStandardData(tc.objectType, tc.ordering)
+
+				foreachTxType(ctx, ds, rev, func(reader datastore.Reader) {
+					// Test that if you ask for resume without an order we error
+					_, err := reader.QueryRelationships(ctx, datastore.RelationshipsFilter{
+						ResourceType: tc.objectType,
+					}, options.WithLimit(&testLimit), options.WithAfter(&core.RelationTuple{}))
+					require.ErrorIs(err, datastore.ErrCursorsWithoutSorting)
+
```

Comment:
I think of the paginated iterator as a helper written from the perspective of the datastore consumer. As such I don't think tests for it belong here. Ideally it would test using mocks. Let me see how long it will take to write that.